Add the fix-dry-run flag

fatfisz opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[X] Add something to the core
[ ] Other, please explain:

**What changes did you make? (Give an overview)**

I've added the new flag, as discussed in #9076.

**Is there anything you'd like reviewers to focus on?**

Nothing in particular, the changes are rather small.

**To do**

- [ ] documentation

platinumazure commented:
Is this cross-platform (i.e., would this work on Windows)?

fatfisz commented:
I'm working on Windows (it works), will check on Ubuntu.

fatfisz commented:
Actually `fs.writeFileSync` did not work as I expected on Ubuntu, but `fs.writeSync` works just fine on both Windows and Ubuntu - I've pushed the changes.

not-an-aardvark commented:
Thanks for the PR. May you please create an issue for this change? From our [pull request guidelines](http://eslint.org/docs/developer-guide/contributing/pull-requests#getting-started), all pull requests that make changes to core require an issue. The idea is that having an issue makes it easier to understand what problem the proposal is trying to solve, and the team can make sure the proposal is the best solution to the problem.

Also, I have a few questions about this approach (we can discuss it further on the issue):

* What does it mean to write to file descriptor 3? How would this typically be used in practice?
* Is the output just a JSON version of the result object? If so, it seems like we don't actually need to send the output to a file descriptor -- we can just remove the restriction on using `--fix` with `--stdin`, and then an integration could use the JSON formatter and handle the output from stdout.

fatfisz commented:
@not-an-aardvark I've created the issue (sorry for not doing this before the PR...).

not-an-aardvark commented:
Removing this from the TSC agenda because the [corresponding issue](https://github.com/eslint/eslint/issues/9076) is also on the TSC agenda and has more up-to-date information.

fatfisz commented:
I've added the `--fix-dry-run` flag instead and updated the PR.

not-an-aardvark commented on lib/cli.js:
```diff
@@ -63,7 +63,7 @@ function translateOptions(cliOptions) {
         cacheFile: cliOptions.cacheFile,
         cacheLocation: cliOptions.cacheLocation,
-        fix: cliOptions.fix && (cliOptions.quiet ? quietFixPredicate : true),
+        fix: (cliOptions.fix || cliOptions.fixDryRun) && (cliOptions.quiet ? quietFixPredicate : true),
```
Can you add a fatal error if `fix` and `fixDryRun` are used together? Currently, `--fix --fix-dry-run` is effectively the same as `--fix`, which is likely to be confusing.

fatfisz replied:
Is there an example of a fatal error on the configuration handling stage that I could refer to? Or should I just throw in `translateOptions`?

not-an-aardvark replied:
I think something like [this](https://github.com/eslint/eslint/blob/439e8e697f761bebeb3dab085ba85139829397b2/lib/cli.js#L175-L176) would be fine.

not-an-aardvark commented on tests/lib/cli.js:
```diff
@@ -858,6 +858,148 @@ describe("cli", () => {
+                results: []
+            });
+            sandbox.stub(fakeCLIEngine.prototype, "getFormatter").returns(() => "done");
+            fakeCLIEngine.outputFixes = sandbox.stub();
```
Should this be `sandbox.mock().never()`?

not-an-aardvark commented on tests/lib/cli.js:
```diff
@@ -858,6 +858,148 @@ describe("cli", () => {
     });
 
+    describe("when passed --fix-dry-run", () => {
+        const sandbox = sinon.sandbox.create();
```
I'm not a big fan of using stubs in tests, since they tend to make the tests fragile ([explanation](https://github.com/airbnb/javascript/blob/f1cc91c88a9feb9d120a61897c2285821afea634/README.md#testing--for-real)). For example, in this case `CLIEngine` makes no guarantees about whether it calls `outputFixes` internally in fix mode -- it only makes guarantees about whether it touches the filesystem. This test makes assertions about whether `outputFixes` is called, so it's possible that it would break or fail to catch bugs later on (since it's testing an implementation detail, rather than the spec).

It looks like this file already contains a bunch of tests with stubs, so I don't really have an objection to adding a few more (they're better than nothing). However, I think it would be good to also add a few integration tests for this behavior in [`tests/bin/eslint.js`](https://github.com/eslint/eslint/blob/ffa021e7696b24722cbbef306d494dd64d7b5215/tests/bin/eslint.js). For example, a test could pipe code to `eslint --stdin --fix-dry-run` and assert that the result contains messages and the output. Another test could run `eslint --fix-dry-run` on a temporary file and assert that the messages/output are correct, but the contents of the file aren't updated.

fatfisz replied:
Ok, I'll add the integration tests then.

not-an-aardvark commented on lib/options.js:
```diff
@@ -191,6 +191,12 @@ module.exports = optionator({
+            default: false,
+            description: "Automatically fix problems without saving the changes to the file system"
+        },
+        {
```
Can you update the [CLI docs](https://github.com/eslint/eslint/blob/ffa021e7696b24722cbbef306d494dd64d7b5215/docs/user-guide/command-line-interface.md) with information on the new flag? There are two places to update:

1. The options summary [here](https://github.com/eslint/eslint/blob/ffa021e7696b24722cbbef306d494dd64d7b5215/docs/user-guide/command-line-interface.md#options) should be updated (I think you can just run `bin/eslint.js -h` on your branch and copy-paste the result).
1. A section should be added [here](https://github.com/eslint/eslint/blob/ffa021e7696b24722cbbef306d494dd64d7b5215/docs/user-guide/command-line-interface.md#--fix). This could describe the flag in a bit more detail, and possibly explain its use cases. For example, it might be worth mentioning that the flag should be used with a formatter that includes the output (such as `json`) if the user/integration wants to access the autofix results. It could also be a good idea to modify the description of `--fix` in the section above to recommend using `--fix-dry-run` for piped-in code.

nzakas commented:
I'd like to suggest changing `--fix-dry-run` to something else as I don't think what this flag does is particularly clear. It's not really a dry run of the fix command, which is what I thought when I first saw this PR, but rather always fixing whatever is passed in to stdin. 

What I'd suggest is changing the `--stdin` flag so it can accept an argument, either `--stdin lint` (which would be the default so `--stdin` still behaves the same way) or `--stdin fix` to have the new behavior of outputting only the fixed version.

not-an-aardvark commented:
@nzakas I'm fine with changing the option name, but I'm not sure making a modifier to `--stdin` is the best way to do it. The proposal for `--fix-dry-run` is to do the same thing as `--fix`, except that the autofix output isn't written to the filesystem. `--fix-dry-run` doesn't actually change the output format -- the autofix output is passed to formatters as the `output` key (which already happens when `--fix` is used), so integrations can use a formatter like `json` to read the autofix output.

Removing the requirement to write output to the filesystem means that it's possible to use `--fix-dry-run` with text piped from `stdin`. However, `--fix-dry-run` doesn't necessarily need to be used with the `--stdin` flag; it can also be used to lint files without writing the autofix results. For example, an integration might plausibly want to lint a file in autofix mode, but then prompt the user for confirmation before writing the autofix output back to the filesystem.

This discussion has gone through a few different proposals, which probably makes it confusing to read. The `--fix-dry-run` option was originally discussed in https://github.com/eslint/eslint/issues/9076#issuecomment-320484888, and I think the ensuing discussion contains a good explanation of how it works.

nzakas commented:
Ah, I see. I totally missed that this isn't just for stdin, so thanks for explaining. 

I withdraw my suggestion for changing the option name.

fatfisz commented:
I'll be fixing the code in a few days, as I'm having some vacation time right now. Correct me if I'm wrong, but there's no rush? ;)

not-an-aardvark commented:
That's right, there's no rush. Enjoy your vacation!

not-an-aardvark commented:
Hi again, just checking in to see if you've had a chance to work on this.

fatfisz commented:
Hi, I came back not too long ago, will take a look soon.

fatfisz commented:
@not-an-aardvark Sorry for the long wait, had to get myself together after the vacation + burn out.

I've fixed everything, squashed, and force-pushed the commit. If you have any other change requests then I'll work on them much faster now ;)

not-an-aardvark commented on docs/user-guide/command-line-interface.md:
```diff
@@ -361,6 +362,16 @@ This option instructs ESLint to try to fix as many issues as possible. The fixes
+
+This option has the same effect as `--fix` with one difference: the fixes are not being saved to the file system. This makes fixing the code from stdin possible.
+
+Because the default formatter does not output the fixed code, you'll have to use another one (e.g. `json`) to get the fixes.
```
I think it would be useful to have an example here, e.g.

```bash
getSomeText | eslint --stdin --fix-dry-run --format=json
```

not-an-aardvark commented on docs/user-guide/command-line-interface.md:
```diff
@@ -361,6 +362,16 @@ This option instructs ESLint to try to fix as many issues as possible. The fixes
+
+#### `--fix-dry-run`
+
+This option has the same effect as `--fix` with one difference: the fixes are not being saved to the file system. This makes fixing the code from stdin possible.
```
Nitpick: I think the word "being" should be removed from the first sentence ("the fixes are not saved to the file system.")

not-an-aardvark replied:
Nitpick: I would modify the second sentence slightly:

```markdown
This makes it possible to fix code from `stdin` (when used with the `--stdin` flag).
```

not-an-aardvark commented on docs/user-guide/command-line-interface.md:
```diff
@@ -361,6 +362,16 @@ This option instructs ESLint to try to fix as many issues as possible. The fixes
+
+Because the default formatter does not output the fixed code, you'll have to use another one (e.g. `json`) to get the fixes.
+
+This flag was added with editor plugins in mind - in some cases they can only use stdio for linting. Without this option enabling autofixing in such tools is close to impossible.
```
I think this paragraph could be rephrased a bit. The main goal of the paragraph is to provide an example use case for the flag, so it doesn't seem useful to describe what editor plugins would be like if this flag didn't exist.

For example:

> This flag can be useful for integrations (e.g. editor plugins) which need to autofix text from the command line without saving it to the filesystem.

not-an-aardvark commented on docs/user-guide/command-line-interface.md:
```diff
@@ -361,6 +362,16 @@ This option instructs ESLint to try to fix as many issues as possible. The fixes
 1. This option throws an error when code is piped to ESLint.
 1. This option has no effect on code that uses processors.
 
+If you want to fix code from stdin or otherwise want to get the fixes without actually writing them to the file, use the [`--fix-dry-run`](--fix-dry-run) option.
```
I think this link is broken -- should it go to `#--fix-dry-run` instead of `--fix-dry-run`?

fatfisz commented:
I fixed the docs based on your suggestions!

not-an-aardvark commented:
We just did a release on Friday, so we're waiting to merge PRs until we determine whether a patch release will be necessary on Monday (see the [release issue](https://github.com/eslint/eslint/issues/9339) for updates). So this will probably get merged sometime around Monday, assuming no one else requests changes on it.

fatfisz commented:
That's awesome, thanks for your time & help!
